framework/src/relationshipmanagers/RedistributeProperties.C

```diff
@@ -0,0 +1,209 @@
+        typedef std::tuple<stored_props_type, dof_id_type, int> stored_elem_type;
+        std::map<processor_id_type, std::vector<stored_elem_type>> props_to_push;
+
+        // Take non-const references here for compatibility with dataStore(T&)
+        for (auto & [elem, elem_props_map] : props_map)
+        {
+          // There better be an element here
+          libmesh_assert(elem);
+
+          // It had better be a *real* element here, not a dangling
+          // pointer
+          libmesh_assert(elem->id() < mesh.max_elem_id());
+          libmesh_assert(elem->processor_id() < mesh.n_processors());
+
+          // If it's not in the particular mesh we're responsible for,
+          // we can skip it.  MOOSE mixes non-displaced with displaced
+          // mesh properties in their storages.
+          if (elem != mesh.query_elem_ptr(elem->id()))
+            continue;
+
+          std::set<processor_id_type> target_pids;
+
+          if (elem->active())
+            target_pids.insert(elem->processor_id());
+          else if (elem->subactive())
+          {
+            libmesh_assert(elem->parent());
+            libmesh_assert(elem->parent()->refinement_flag() == Elem::JUST_COARSENED);
+            target_pids.insert(elem->parent()->processor_id());
+          }
+          else
+          {
+            libmesh_assert(elem->ancestor());
+            libmesh_assert(elem->has_children());
+            for (const Elem & child : elem->child_ref_range())
+              target_pids.insert(child.processor_id());
+          }
+
+          // If we're being migrated elsewhere and we're not needed
+          // for AMR/C locally then we should erase the local entry
+          bool remote_pid = false, local_pid = false;
+          for (processor_id_type target_pid : target_pids)
+            if (target_pid == pid)
+              local_pid = true;
+            else
+            {
+              remote_pid = true;
+
+              stored_props_type stored_props;
+              unsigned int n_q_points = 0;
+              for (auto & [prop_id, prop_vals] : elem_props_map)
+              {
+                libmesh_assert(!prop_vals.empty());
+
+                for (PropertyValue * prop : prop_vals)
+                {
+                  libmesh_assert(prop);
+                  n_q_points = std::max(n_q_points, prop->size());
+                }
+
+                std::ostringstream oss;
+                dataStore(oss, prop_vals, nullptr);
+                stored_props[prop_id] = oss.str();
+              }
+
+              // Get the stored data ready to push to the element's new
+              // processor
+              props_to_push[target_pid].emplace_back(
+                  std::move(stored_props), elem->id(), n_q_points);
+            }
+
+          if (remote_pid && !local_pid)
+            elems_being_migrated.insert(elem);
+        }
+
+        auto recv_functor = [&, mat_prop_store_ptr = mat_prop_store](
```

Comment:
I've never seen this before. This performs a copy I take it?

Reply:
Yup.  Fortunately a pointer copy.

There was a big rat hole behind this: https://stackoverflow.com/a/46115028